community/kernel/src/main/java/org/neo4j/kernel/monitoring/Monitors.java

```diff
@@ -71,155 +74,81 @@ public Monitors( LogProvider logProvider )
         return monitorClass.cast( Proxy.newProxyInstance( classLoader, new Class<?>[]{monitorClass}, monitorInvocationHandler ) );
     }
 
-    public synchronized void addMonitorListener( final Object monitorListener, String... tags )
+    public synchronized void addMonitorListener( Object monitorListener, String... tags )
     {
-        MonitorListenerInvocationHandler monitorListenerInvocationHandler =
-                tags.length == 0 ? new UntaggedMonitorListenerInvocationHandler( monitorListener )
-                                 : new TaggedMonitorListenerInvocationHandler( monitorListener, tags );
+        MonitorListenerInvocationHandler monitorListenerInvocationHandler = createInvocationHandler( monitorListener, tags );
 
-        for ( Class<?> monitorInterface : getInterfacesOf( monitorListener.getClass() ) )
+        List<Class<?>> listenerInterfaces = getAllInterfaces( monitorListener );
+        methodsStream( listenerInterfaces ).forEach( method ->
         {
-            for ( final Method method : monitorInterface.getMethods() )
-            {
-                monitorListeners.put(
-                        Predicate.isEqual( method ),
-                        monitorListenerInvocationHandler );
-
-                recalculateMethodListeners( method );
-            }
-        }
+            Set<MonitorListenerInvocationHandler> methodHandlers =
+                    methodMonitorListeners.computeIfAbsent( method, f -> Collections.newSetFromMap( new ConcurrentHashMap<>() ) );
+            methodHandlers.add( monitorListenerInvocationHandler );
+        } );
+        monitoredInterfaces.addAll( listenerInterfaces );
     }
 
     public synchronized void removeMonitorListener( Object monitorListener )
     {
-        Iterator<Map.Entry<Predicate<Method>, MonitorListenerInvocationHandler>> iter =
-                monitorListeners.entrySet().iterator();
-
-        while ( iter.hasNext() )
-        {
-            Map.Entry<Predicate<Method>, MonitorListenerInvocationHandler> handlerEntry = iter.next();
-            if ( handlerEntry.getValue() instanceof UntaggedMonitorListenerInvocationHandler )
+        List<Class<?>> listenerInterfaces = getAllInterfaces( monitorListener );
+        methodsStream( listenerInterfaces ).forEach( method -> {
+            Set<MonitorListenerInvocationHandler> handlers = methodMonitorListeners.get( method );
+            handlers.removeIf( handler -> monitorListener.equals( handler.getMonitorListener() ) );
+            if ( handlers.isEmpty() )
             {
-                UntaggedMonitorListenerInvocationHandler handler =
-                        (UntaggedMonitorListenerInvocationHandler) handlerEntry.getValue();
-
-                if ( handler.getMonitorListener() == monitorListener )
-                {
-                    iter.remove();
-                }
+                methodMonitorListeners.remove( method );
             }
```

Comment:
You can use `computeIfPresent` on the `methodMonitorListeners` map, to do this entire modification atomically in the `ConcurrentHashMap`.

Then you'll only need to synchronise around the `monitoredInterfaces` collection.

I think that might speed up things even more.